stdlib/Printf/src/Printf.jl

```diff
@@ -61,9 +342,202 @@ macro sprintf(args...)
+    letexpr = printf_expr("@sprintf", __source__, :(IOBuffer()), args[1], args[2:end])
     push!(letexpr.args[2].args, :(String(take!(out))))
     letexpr
 end
+"""
+    format" literal format string"
+
+Generate a format function f, which can be called like f(io, args...).
+That has the same effect as `@printf(io, "format string", args..)` while the generated
+code is not inlined, but in a separate local function.
+The format function is generated only once for each different format string.
+"""
+macro format_str(arg)
+    format_expr(__source__, arg)
+end
+
+##### formatting functions
+## kind of type intersection, base on the symbols :Any, :Real, :Integer, :Ptr
+# note: gen_d returns :Real (not :Integer)
+#       gen_c returns :Integer (not :Char)
+#
+function typeinter(a::Symbol, b::Symbol)
+    a === :Any ? b :
+    b === :Any ? a :
+    a === b ?    a :
+    a === :Real ? b :
+    b === :Real ? a :
+                :PIC
+end
+
+## map to the actual argument types according to format specs
+#
+function typearg(a::Symbol)
+    a === :Any ? a :
+    a === :Real ? a :
+    a === :Integer ? Expr(:curly, :Union, :Char, a) :
+    a === :Ptr ? Expr(:curly, :Union, :Integer, a) :
+                :Integer
+end
+
+## for all arguments determine the formal argument types
+# to be accepted by the function
+# Re-use the varaible names of the symbolic arguments
+# If multiple format specs are delivered by the same argument
+# the argument type must be restricted reasonably
+#
+function typeconsolidate(sargs, perm)
+    n = max(0, perm...)
+    m = length(perm)
+    fargs = Vector{Any}(nothing, n)
+
+    for i = 1:m
+        j = perm[i]
+        faj = fargs[j]
+        if faj === nothing
+            fargs[j] = sargs[i] # :(var::type)
+        else
+            ft = faj.args[2]
+            st = sargs[i].args[2]
+            if st != :Any
+                if ft == :Any
+                    faj.args[2] = st
+                else
+                    faj.args[2] = typeinter(ft, st)
+                end
+            end
+        end
+    end
+    for j = 1:n
+        faj = fargs[j]
+        if faj === nothing
+            fargs[j] = :(::Any) # unused argument without variable name
+        else
+            faj.args[2] = typearg(faj.args[2]) # final type
+        end
+    end
+    fargs
+end
+
+
+## generate function definition for format function
+#
+function format_expr(line::LineNumberNode, fmt::AbstractString)
+    sargs, blk, perm = gen(fmt)
+    fargs = typeconsolidate(sargs, perm)
+    args = map(p->p.args[1], fargs)
+    blk = bind_vars!("format", false, :out, sargs, blk, perm, args)
+
+    fun = :($(genfun())(out::IO) = nothing)
+    append!(fun.args[1].args, fargs)
+    fun.args[2].args[2] = blk
+    fun.args[2].args[1] = line
+    fun
+end
+
+# generate anonymous function
+function format_fun(line::LineNumberNode, fmt)
+    f(io::IO, args...) = Base.invokelatest(eval(format_expr(line, fmt)), io, args...)
```

Comment:
The `eval(format_expr(...` call should be moved outside this function so it's only done once per format.

Reply:
This format_fun is only called once per format string at line 474 in the get-do body of function `format`. That is the reason for the global dictionary `ALL_FORMATS`.
All calls to `printf` and `sprintf` call in to `format`. The macro implementation calls `format_expr` directly during macro expansion.

Reply:
Yes, but the returned function `f` calls `eval(format_expr(...` every time it's called.

Reply:
You are right! That was a stupid optimization of the last minute

Reply:
That reminds me, I had a hard time to accept, why not simply 
```
format_fun(line, fmt) = eval(format_expr(line, fmt)
``` 
 was possible, and I have to call `invokelatest`. Still bothering..., I had hoped for an improvement here: [discourse](https://discourse.julialang.org/t/help-calling-a-function-defined-from-expressions/1340/7)